util/testutil/port.go

```diff
@@ -15,21 +15,56 @@ package testutil
 
 import (
 	"net"
+	"sync"
 	"testing"
 )
 
+var (
+	mu        sync.Mutex
+	usedPorts []int
+)
+
 // RandomUnprivilegedPort returns valid unprivileged random port number which can be used for testing.
 func RandomUnprivilegedPort(t *testing.T) int {
 	t.Helper()
+	mu.Lock()
+
+	port, err := getPort()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for portWasUsed(port) {
+		port, err = getPort()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	usedPorts = append(usedPorts, port)
+	mu.Unlock()
```

Comment:
Let's do

```go
mu.Lock()
defer mu.Unlock()
```

at the top, that way we have to think less about unlocking correctness.